landlab/utils/flow__distance.py

```diff
@@ -0,0 +1,212 @@
+        
+        Now, let's change to D4 the flow_director method, that does not 
+        consider diagonal links bewtween nodes.
+        
+    >>> from landlab import RasterModelGrid
+    >>> from landlab.components import FlowAccumulator
+    >>> from landlab.utils.flow__distance import calculate_flow__distance
+    >>> mg = RasterModelGrid((5, 4), spacing=(1, 1))
+    >>> elev = np.array([0.,  0.,  0., 0.,
+        ...              0., 21., 10., 0.,
+        ...              0., 31., 20., 0.,
+        ...              0., 32., 30., 0.,
+        ...              0.,  0.,  0., 0.])
+    >>> mg.add_field('node','topographic__elevation', z)
+    >>> mg.set_closed_boundaries_at_grid_edges(bottom_is_closed=True, left_is_closed=True, right_is_closed=True, top_is_closed=True)
+    >>> fr = FlowAccumulator(mg, flow_director = 'D4')
+    >>> fr.run_one_step()
+    >>> flow_distance = calculate_flow__distance(mg, add_to_grid=True, noclobber=False)
+    >>> #Set to zero the flow distance values of boundary nodes and watershed's outlet
+    >>> flow_distance[mg.boundary_nodes] = 0
+    >>> outlet_id = 6
+    >>> flow_distance[outlet_id] = 0
+    >>> mg.at_node['flow_distance']
+    >>> array([  0,  0,  0,  0,
+                 0,  1,  0,  0,
+                 0,  2,  1, 0,
+                 0, 3, 2, 0,
+                 0, 0, 0, 0])
+                 
+        The flow__distance utility can also work on irregular grids. For the example we 
+        will use a Hexagonal Model Grid, a special type of Voroni Grid that has 
+        regularly spaced hexagonal cells. We will also set the dx spacing such 
+        that each cell has an area of one.  
+        
+    >>> from landlab import HexModelGrid, CLOSED_BOUNDARY
+    >>> from landlab.components import FlowAccumulator
+    >>> from landlab.utils.flow__distance import calculate_flow__distance
+    >>> dx=(2./(3.**0.5))**0.5
+    >>> mg = HexModelGrid(5,3, dx)
+    >>> mg.add_field('topographic__elevation', mg.node_x + np.round(mg.node_y), at = 'node')
+    >>> hmg.status_at_node[0] = CLOSED_BOUNDARY
+    >>> fr = FlowAccumulator(hmg, flow_director = 'D4')
+    >>> fr.run_one_step()
+    >>> flow_distance = calculate_flow__distance(hmg, add_to_grid=True, noclobber=False)
+    >>> #Even in this case, boundary and outlet nodes should have a flow distance equal to zero
+    >>> flow_distance[hmg.boundary_nodes] = 0
+    >>> outlet_id = 4
+    >>> flow_distance[outlet_id] = 0
+    
+        
+    """
+    # check that flow__receiver nodes exists
+    if 'flow__receiver_node' not in grid.at_node:
+        raise FieldError("A 'flow__receiver_node' field is required at the "
+                         "nodes of the input grid.")
+    if 'flow__upstream_node_order' not in grid.at_node:
+        raise FieldError("A 'flow__upstream_node_order' field is required at the "
+                         "nodes of the input grid.")
+
+    # get an array of grid nodes
+    grid_nodes = grid.nodes.flatten()
+
```

Comment:
@Giuseppecipolla95 
I think we are about ready to merge this, nice work!

I think that we can delete these two lines since `grid_nodes` is no longer used later. 

```python
# get an array of grid nodes
grid_nodes = grid.nodes.flatten()
``` 

You can either delete the lines yourself or add me as a collaborator on this fork. Either option is fine with me. 

Once that is done, we'll merge this branch and @nathanlyons will follow up with you regarding finishing the tutorial.

Reply:
@kbarnhart I've added you as a collaborator on the fork, so you should be able to delete the two lines indicated by yourself on the comment.
Regarding the tutorial, I think it should be complete. 
@nathanlyons Please confirm it or let me know if anything else should be fixed.